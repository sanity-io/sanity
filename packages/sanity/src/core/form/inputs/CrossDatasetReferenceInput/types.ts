import {PreviewValue} from '@sanity/types'
import {Observable} from 'rxjs'
import {DocumentAvailability} from '../../../preview'

/** @internal */
export interface CrossDatasetReferenceInfo {
  id: string
  type: string | undefined
  availability: DocumentAvailability | null
  preview: {
    published: PreviewValue | undefined
  }
}

/** @internal */
export interface SearchState {
  hits: CrossDatasetSearchHit[]
  searchString?: string
  isLoading: boolean
}

/** @internal */
export type CrossDatasetSearchFunction = (query: string) => Observable<CrossDatasetSearchHit[]>

/** @internal */
export interface CrossDatasetSearchHit {
  id: string
  type: string
  published: undefined | {_id: string; _type: string}
}
