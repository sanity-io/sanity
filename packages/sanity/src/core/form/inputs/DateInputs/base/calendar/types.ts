export interface CalendarLabels {
  goToTomorrow: string
  goToToday: string
  goToYesterday: string
  goToPreviousYear: string
  goToNextYear: string
  goToPreviousMonth: string
  goToNextMonth: string
  selectHour: string
  selectMinute: string
  setToCurrentTime: string
  monthNames: MonthNames
  weekDayNamesShort: WeekDayNames
  setToTimePreset: (time: string, date: Date) => string
}

export type WeekDayNames = [
  mon: string,
  tue: string,
  wed: string,
  thu: string,
  fri: string,
  sat: string,
  sun: string,
]

export type MonthNames = [
  jan: string,
  feb: string,
  mar: string,
  apr: string,
  may: string,
  jun: string,
  jul: string,
  aug: string,
  sep: string,
  oct: string,
  nov: string,
  dec: string,
]
