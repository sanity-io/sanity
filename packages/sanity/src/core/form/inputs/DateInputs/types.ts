export type ParseResult = {isValid: boolean; date?: Date; error?: string} & (
  | {isValid: true; date: Date}
  | {isValid: false; error?: string}
)
