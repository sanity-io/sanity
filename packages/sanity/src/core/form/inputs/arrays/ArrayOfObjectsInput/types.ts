/* eslint-disable camelcase */

import type {ComponentType} from 'react'
import {KeyedSegment} from '@sanity/types'

export type ModalType = 'modal' | 'fullscreen' | string

/** @public */
export type ArrayInputMember = {
  _type?: string
  _key: string
  [key: string]: any
}

/** @public */
export interface InsertEvent {
  item: ArrayInputMember
  position: 'before' | 'after'
  referenceItem: KeyedSegment | number
  edit?: boolean
}

/** @internal */
export type ReferenceItemComponentType = ComponentType<any>
