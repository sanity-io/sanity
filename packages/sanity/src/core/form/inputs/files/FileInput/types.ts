import {type AssetFromSource, type AssetSource} from '@sanity/types'

import {type FileInfo} from '../common/styles'
import {type BaseFileInputProps} from './FileInput'

export interface FileAssetProps extends Omit<BaseFileInputProps, 'renderDefault'> {
  browseButtonElementRef: React.RefObject<HTMLButtonElement | null>
  clearField: () => void
  hoveringFiles: FileInfo[]
  isBrowseMenuOpen: boolean
  isStale: boolean
  isUploading: boolean
  onCancelUpload?: () => void
  onClearUploadStatus: () => void
  onSelectAssets: (assetsFromSource: AssetFromSource[]) => void
  onSelectFiles: (assetSource: AssetSource, files: File[]) => void
  onStale: () => void
  selectedAssetSource: AssetSource | null
  setBrowseButtonElement: (element: HTMLButtonElement | null) => void
  setHoveringFiles: (hoveringFiles: FileInfo[]) => void
  setIsBrowseMenuOpen: (isBrowseMenuOpen: boolean) => void
  setIsUploading: (isUploading: boolean) => void
  setSelectedAssetSource: (assetSource: AssetSource | null) => void
}
