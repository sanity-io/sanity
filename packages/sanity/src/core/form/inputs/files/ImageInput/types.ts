import {type SanityClient} from '@sanity/client'
import {
  type AssetSource,
  type Image as BaseImage,
  type ImageAsset,
  type ImageSchemaType,
  type UploadState,
} from '@sanity/types'
import {type Observable} from 'rxjs'

import {type UploaderResolver} from '../../../studio/uploads/types'
import {type ObjectInputProps} from '../../../types'
import {type ImageUrlBuilder} from '../types'

/**
 * @hidden
 * @internal
 */
export type FileInfo = {
  type: string // mime type
  kind: string // 'file' or 'string'
}

/**
 * @hidden
 * @beta
 */
export interface BaseImageInputValue extends Partial<BaseImage> {
  _upload?: UploadState
}

/**
 * @hidden
 * @beta */
export interface BaseImageInputProps
  extends ObjectInputProps<BaseImageInputValue, ImageSchemaType> {
  assetSources: AssetSource[]
  directUploads?: boolean
  imageUrlBuilder: ImageUrlBuilder
  observeAsset: (documentId: string) => Observable<ImageAsset>
  resolveUploader: UploaderResolver
  client: SanityClient
  t: (key: string, values?: Record<string, string>) => string
}
