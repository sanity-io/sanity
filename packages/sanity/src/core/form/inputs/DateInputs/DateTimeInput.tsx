import {format, parse} from '@sanity/util/legacyDateFormat'
import {getMinutes, setMinutes, parseISO} from 'date-fns'
import React, {useCallback} from 'react'
import {set, unset} from '../../patch'
import {StringInputProps} from '../../types'
import {CommonDateTimeInput} from './CommonDateTimeInput'
import {ParseResult} from './types'
import {isValidDate} from './utils'

interface ParsedOptions {
  dateFormat: string
  timeFormat: string
  timeStep: number
  calendarTodayLabel: string
}

interface SchemaOptions {
  dateFormat?: string
  timeFormat?: string
  timeStep?: number
  calendarTodayLabel?: string
}

/**
 * @hidden
 * @beta */
export type DateTimeInputProps = StringInputProps

const DEFAULT_DATE_FORMAT = 'YYYY-MM-DD'
const DEFAULT_TIME_FORMAT = 'HH:mm'

function parseOptions(options: SchemaOptions = {}): ParsedOptions {
  return {
    dateFormat: options.dateFormat || DEFAULT_DATE_FORMAT,
    timeFormat: options.timeFormat || DEFAULT_TIME_FORMAT,
    timeStep: ('timeStep' in options && Number(options.timeStep)) || 1,
    calendarTodayLabel: options.calendarTodayLabel || 'Today',
  }
}

function serialize(date: Date) {
  return date.toISOString()
}
function deserialize(isoString: string): ParseResult {
  const deserialized = new Date(isoString)
  if (isValidDate(deserialized)) {
    return {isValid: true, date: deserialized}
  }
  return {isValid: false, error: `Invalid date value: "${isoString}"`}
}

// enforceTimeStep takes a dateString and datetime schema options and enforces the time
// to be within the configured timeStep
function enforceTimeStep(dateString: string, timeStep: number) {
  if (!timeStep || timeStep === 1) {
    return dateString
  }

  const date = parseISO(dateString)
  const minutes = getMinutes(date)
  const leftOver = minutes % timeStep
  if (leftOver !== 0) {
    return serialize(setMinutes(date, minutes - leftOver))
  }

  return serialize(date)
}

/**
 * @hidden
 * @beta */
export function DateTimeInput(props: DateTimeInputProps) {
  const {onChange, schemaType, value, elementProps} = props

  const {dateFormat, timeFormat, timeStep} = parseOptions(schemaType.options)

  const handleChange = useCallback(
    (nextDate: string | null) => {
      let date = nextDate
      if (date !== null && timeStep > 1) {
        date = enforceTimeStep(date, timeStep)
      }

      onChange(date === null ? unset() : set(date))
    },
    [onChange, timeStep],
  )

  const formatInputValue = React.useCallback(
    (date: Date) => format(date, `${dateFormat} ${timeFormat}`),
    [dateFormat, timeFormat],
  )

  const parseInputValue = React.useCallback(
    (inputValue: string) => parse(inputValue, `${dateFormat} ${timeFormat}`),
    [dateFormat, timeFormat],
  )

  return (
    <CommonDateTimeInput
      {...elementProps}
      onChange={handleChange}
      deserialize={deserialize}
      formatInputValue={formatInputValue}
      parseInputValue={parseInputValue}
      placeholder={schemaType.placeholder}
      selectTime
      serialize={serialize}
      timeStep={timeStep}
      value={value}
    />
  )
}
