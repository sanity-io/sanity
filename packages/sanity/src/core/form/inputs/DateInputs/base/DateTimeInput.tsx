import React, {forwardRef, useCallback, useRef, useState} from 'react'
import FocusLock from 'react-focus-lock'
import {Box, Button, LayerProvider, Popover, useClickOutside, useForwardedRef} from '@sanity/ui'
import {CalendarIcon} from '@sanity/icons'
import {DatePicker} from './DatePicker'
import {LazyTextInput} from './LazyTextInput'

export interface DateTimeInputProps {
  customValidity?: string
  id?: string
  inputValue?: string
  onChange: (date: Date | null) => void
  onInputChange?: (event: React.FocusEvent<HTMLInputElement>) => void
  placeholder?: string
  readOnly?: boolean
  selectTime?: boolean
  timeStep?: number
  value?: Date
}

export const DateTimeInput = forwardRef(function DateTimeInput(
  props: DateTimeInputProps,
  ref: React.ForwardedRef<HTMLInputElement>,
) {
  const {value, inputValue, onInputChange, onChange, selectTime, timeStep, ...rest} = props
  const [popoverRef, setPopoverRef] = useState<HTMLElement | null>(null)
  const forwardedRef = useForwardedRef(ref)
  const buttonRef = useRef(null)

  const [isPickerOpen, setPickerOpen] = useState(false)

  useClickOutside(() => setPickerOpen(false), [popoverRef])

  const handleDeactivation = useCallback(() => {
    forwardedRef.current?.focus()
    forwardedRef.current?.select()
  }, [forwardedRef])

  const handleKeyUp = useCallback((e: any) => {
    if (e.key === 'Escape') {
      setPickerOpen(false)
    }
  }, [])

  const handleClick = useCallback(() => setPickerOpen(true), [])

  const suffix = (
    <Box padding={1}>
      <Button
        ref={buttonRef}
        icon={CalendarIcon}
        mode="bleed"
        padding={2}
        onClick={handleClick}
        style={{display: 'block'}}
        data-testid="select-date-button"
      />
    </Box>
  )

  return (
    <LazyTextInput
      ref={forwardedRef}
      {...rest}
      value={inputValue}
      onChange={onInputChange}
      suffix={
        isPickerOpen ? (
          // Note: we're conditionally inserting the popover here due to an
          // issue with popovers rendering incorrectly on subsequent renders
          // see https://github.com/sanity-io/design/issues/519
          <LayerProvider zOffset={1000}>
            <Popover
              constrainSize
              data-testid="date-input-dialog"
              portal
              content={
                <Box overflow="auto">
                  <FocusLock onDeactivation={handleDeactivation}>
                    <DatePicker
                      selectTime={selectTime}
                      timeStep={timeStep}
                      onKeyUp={handleKeyUp}
                      value={value}
                      onChange={onChange}
                    />
                  </FocusLock>
                </Box>
              }
              open
              placement="bottom"
              ref={setPopoverRef}
              radius={2}
            >
              {suffix}
            </Popover>
          </LayerProvider>
        ) : (
          suffix
        )
      }
    />
  )
})
