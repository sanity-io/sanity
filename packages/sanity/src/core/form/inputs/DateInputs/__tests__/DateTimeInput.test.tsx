import {defineField} from '@sanity/types'
import {fireEvent} from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import React from 'react'
import {renderStringInput} from '../../../../../../test/form'
import {DateTimeInput} from '../DateTimeInput'

// NOTE: for the tests to be deterministic we need this to ensure tests are run in a predefined timezone
// see globalSetup in jest config for details about how this is set up
test('timezone for the test environment should be set to America/Los_Angeles', () => {
  expect(Intl.DateTimeFormat().resolvedOptions().timeZone).toBe('America/Los_Angeles')
})

test('does not emit onChange after invalid value has been typed', async () => {
  const {onChange, result} = await renderStringInput({
    fieldDefinition: defineField({
      type: 'datetime',
      name: 'test',
    }),
    render: (inputProps) => <DateTimeInput {...inputProps} />,
  })

  const input = result.container.querySelector('input')!

  userEvent.type(input, 'this is invalid')
  expect(input.value).toBe('this is invalid')
  expect(onChange.mock.calls.length).toBe(0)

  fireEvent.blur(input)

  expect(onChange.mock.calls.length).toBe(0)
})

test('emits onChange on correct format if a valid value has been typed', async () => {
  const {onChange, result} = await renderStringInput({
    fieldDefinition: defineField({
      type: 'datetime',
      name: 'test',
    }),
    render: (inputProps) => <DateTimeInput {...inputProps} />,
  })

  const input = result.container.querySelector('input')!

  // NOTE: the date is entered and displayed in local timezone
  // (which is hardcoded to America/Los_Angeles)
  userEvent.type(input, '2021-03-28 10:23')
  expect(input.value).toBe('2021-03-28 10:23')

  fireEvent.blur(input)

  // NOTE: the date is entered and displayed in local timezone but stored in utc
  expect(onChange.mock.calls).toMatchSnapshot()
})

test('formatting of deserialized value', async () => {
  const {result} = await renderStringInput({
    fieldDefinition: defineField({
      type: 'datetime',
      name: 'test',
    }),
    props: {documentValue: {test: '2021-03-28T17:23:00.000Z'}},
    render: (inputProps) => <DateTimeInput {...inputProps} />,
  })

  const input = result.container.querySelector('input')!

  // const {textInput} = renderInput({value: '2021-03-28T17:23:00.000Z'} as any)
  expect(input.value).toBe('2021-03-28 10:23')
})
