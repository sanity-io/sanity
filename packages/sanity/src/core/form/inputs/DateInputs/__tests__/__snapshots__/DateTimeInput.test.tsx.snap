// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`emits onChange on correct format if a valid value has been typed 1`] = `
[
  [
    {
      "patchType": Symbol(sanity.patch),
      "path": [],
      "type": "set",
      "value": "2021-03-28T17:23:00.000Z",
    },
  ],
]
`;
